src/analyses/does_remove_const.cpp

```diff
@@ -0,0 +1,166 @@
+/*******************************************************************\
+
+ Module: Analyses
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#include <goto-programs/goto_program.h>
+#include <util/type.h>
+#include <util/expr.h>
+#include <util/std_code.h>
+#include <util/base_type.h>
+#include <ansi-c/c_qualifiers.h>
+
+#include "does_remove_const.h"
+
+/*******************************************************************\
+
+Function: does_remove_constt::does_remove_constt
+
+  Inputs:
+   goto_program - the goto program to check
+   ns - the namespace of the goto program (used for checking type equality)
+
+ Outputs:
+
+ Purpose: A naive analysis to look for casts that remove const-ness from
+          pointers.
+
+\*******************************************************************/
+
+does_remove_constt::does_remove_constt(
+  const goto_programt &goto_program,
+  const namespacet &ns):
+    goto_program(goto_program),
+    ns(ns)
+{}
+
+/*******************************************************************\
+
+Function: does_remove_constt::operator()
+
+  Inputs:
+
+ Outputs: Returns true if the program contains a const-removing cast
+
+ Purpose: A naive analysis to look for casts that remove const-ness from
+          pointers.
+
+\*******************************************************************/
+
+bool does_remove_constt::operator()() const
+{
+  for(const goto_programt::instructiont &instruction :
+    goto_program.instructions)
+  {
+    if(instruction.is_assign())
+    {
+      const code_assignt assign=to_code_assign(instruction.code);
+      const typet &rhs_type=assign.rhs().type();
+      const typet &lhs_type=assign.lhs().type();
+
+      // Compare the types recursively for a point where the rhs is more
+      // const that the lhs
+      if(!is_type_at_least_as_const_as(lhs_type, rhs_type))
+      {
+        return true;
+      }
+
+      bool sub_expr_lose_const=does_expr_lose_const(assign.rhs());
+      if(sub_expr_lose_const)
+      {
+        return true;
+      }
+    }
```

Comment:
I'm always a bit scared when I see too many closing parentheses, so here is what *I* would have done. This is *not* a request for changes, but just my personal style: 1) Use `if(!instruction.is_assign()) continue;` above. 2) Drop the braces around `return true;` 3) Not use a temporary to store the return value.

Only make changes if you share any of those feelings.